### ISSUE TITLE
Expect intrinsics is silently dropped

### ISSUE THREAD

**Issue by asl** [llvm/llvm-project#56129]
After upgrading some legacy code, I'm seeing that `llvm.expect` intrinsic is silently dropped. This seems to be an outcome of fixing https://github.com/llvm/llvm-project/issues/48680 fix. What happens is that the pass is unable to understand the more complex patterns in unoptimized code.

Here is the typical example (unoptimized code from the frontend):
```
...
  %8 = or i4 0, %k.479.476
  %eq12 = icmp eq i4 %8, -8
  %and13 = and i1 %and, %eq12
  %or14 = or i1 %or9, %and13
  store i1 %or14, i1* %b_0, align 1
  %9 = load i1, i1* %b_0, align 1
  %10 = call i1 @llvm.expect.i1(i1 %9, i1 false)
  store i1 %10, i1* %tmp, align 1
  %11 = load i1, i1* %tmp, align 1
  br i1 %11, label %.if.true.0, label %.if.false.1
```

This pattern is not understood by `LowerExpect` pass and therefore the intrinsic is silently dropped. Before https://github.com/llvm/llvm-project/commit/661cc71a1c50081389d73b2baae02f51df670ba1 we got `SROA` run before `LowerExpect` and therefore the pass saw fully supported:
```
  %8 = or i4 0, %k.479.476
  %eq12 = icmp eq i4 %8, -8
  %and13 = and i1 %and, %eq12
  %or14 = or i1 %or9, %and13
  %9 = call i1 @llvm.expect.i1(i1 %or14, i1 false)
  br i1 %9, label %.if.true.0, label %.if.false.1
```

Tagging @rotateright

**Comment by asl** [comment]
@nikic Right. Looks like a clear regression to me. As now frontends cannot emit arbitrary non-optimized code and expect that optimization passes would clean the things up. Instead they need to emit something that would be consumed / expected by some optimization passes (`LowerExpect` as of now). Surely it worked before because some early optimization passes made the code mode accessible for pattern-matching and because clang emits some particular sequence of code (though it may stop at some point easily), but overall area of applicability of this intrinsic now is much lower than it used to be for no particular reason IMO.

In any case, it's good to have issue documenting the problem. Instead of some comments in closed PR and some Rust issues. I had to revert D100213 locally as it caused lots of performance regressions in well-annotated user code.

**Comment by davidbolvansky** [comment]
Given that patch caused so many problems, it should be reverted.
